File: js/packages/trellis/contract_support/protocol_artifacts.ts
import { digestJson, isJsonValue, type JsonValue } from "./canonical.ts";
import { CONTRACT_RUNTIME, type ContractRuntime } from "./contract_runtime.ts";
import { actionSource } from "./descriptors.ts";

type JsonObject = Record<string, JsonValue>;

/** Canonical protocol artifacts compiled from a TypeScript contract manifest. */
export type CompiledProtocolArtifacts = {
  api: JsonObject;
  participant: JsonObject;
  referencedApis: readonly JsonObject[];
};

function object(value: JsonValue | undefined): JsonObject | undefined {
  return value !== null && typeof value === "object" && !Array.isArray(value)
    ? value as JsonObject
    : undefined;
}

function checkedObject(value: Readonly<Record<string, unknown>>): JsonObject {
  if (!Object.values(value).every(isJsonValue)) {
    throw new Error("Contract manifest must contain only JSON values");
  }
  return value as JsonObject;
}

type ContractInput = Readonly<Record<string, unknown>> & {
  readonly CONTRACT?: Readonly<Record<string, unknown>>;
  readonly CONTRACT_DIGEST?: string;
  readonly [CONTRACT_RUNTIME]?: ContractRuntime;
};

function sourceManifest(contract: ContractInput): JsonObject {
  return checkedObject(contract.CONTRACT ?? contract);
}

async function compileReferencedApi(
  source: Readonly<Record<string, unknown>> | {
    readonly artifact: Readonly<Record<string, unknown>>;
    readonly digest: string;
  },
): Promise<JsonObject> {
  const artifact = Reflect.get(source, "artifact");
  const digest = Reflect.get(source, "digest");
  const wrapped = artifact !== null && typeof artifact === "object" &&
    !Array.isArray(artifact) && typeof digest === "string";
  const manifest = checkedObject(
    wrapped ? artifact as Readonly<Record<string, unknown>> : source,
  );
  if (manifest.format === "trellis.api.v1") {
    return structuredClone(manifest);
  }
  return await compileApi(manifest, wrapped ? digest as string : undefined);
}

function normalizeSchema(value: JsonValue): void {
  if (Array.isArray(value)) {
    value.forEach(normalizeSchema);
    return;
  }
  const record = object(value);
  if (!record) return;
  if ("patternProperties" in record) {
    delete record.patternProperties;
    record.additionalProperties = true;
  }
  Object.values(record).forEach(normalizeSchema);
}

function copy(source: JsonObject, target: JsonObject, key: string): void {
  const value = source[key];
  const record = object(value);
  if (
    value !== undefined && value !== null &&
    !(record && Object.keys(record).length === 0)
  ) target[key] = structuredClone(value);
}

async function compileApi(
  contract: JsonObject,
  contractDigest?: string,
): Promise<JsonObject> {
  const api: JsonObject = { format: "trellis.api.v1" };
  for (
    const field of [
      "id",
      "displayName",
      "description",
      "docs",
      "schemas",
      "exports",
    ]
  ) copy(contract, api, field);
  Object.values(object(api.schemas) ?? {}).forEach(normalizeSchema);
  const schemas = object(api.schemas) ?? {};
  if ("TrellisContractArtifactIdentity" in schemas) {
    throw new Error("TrellisContractArtifactIdentity is reserved");
  }
  schemas.TrellisContractArtifactIdentity = {
    type: "string",
    const: contractDigest ?? (await digestJson(contract)).digest,
  };
  api.schemas = schemas;

  const capabilityAllows: Record<string, JsonValue[]> = {};
  const addCapability = (
    capability: JsonValue,
    action: string,
    target: JsonObject,
  ) => {
    if (typeof capability !== "string") return;
    (capabilityAllows[capability] ??= []).push({ action, target });
  };
  const apiId = String(contract.id);
  for (
    const [section, actionMap] of [
      ["rpc", { call: "call" }],
      ["operations", {
        call: "invoke",
        observe: "observe",
        cancel: "cancel",
      }],
      ["events", { publish: "publish", subscribe: "subscribe" }],
      ["feeds", { subscribe: "subscribe" }],
    ] as const
  ) {
    for (
      const [name, value] of Object.entries(object(contract[section]) ?? {})
        .sort(([left], [right]) => left.localeCompare(right))
    ) {
      const definition = object(value);
      const capabilities = object(definition?.capabilities);
      for (const [direction, action] of Object.entries(actionMap)) {
        for (
          const capability of capabilities?.[direction] as JsonValue[] ?? []
        ) {
          addCapability(capability, action, {
            kind: "apiSurface",
            api: apiId,
            surface: section === "operations"
              ? "operation"
              : section === "events"
              ? "event"
              : section === "feeds"
              ? "feed"
              : "rpc",
            name,
          });
        }
      }
      if (section === "operations") {
        for (const capability of capabilities?.control as JsonValue[] ?? []) {
          for (
            const signal of Object.keys(object(definition?.signals) ?? {})
              .sort()
          ) {
            addCapability(capability, "control", {
              kind: "operationSignal",
              api: apiId,
              operation: name,
              signal,
            });
          }
        }
      }
    }
  }
  if (Object.keys(capabilityAllows).length > 0) {
    api.capabilities = Object.fromEntries(
      Object.entries(capabilityAllows).sort(([left], [right]) =>
        left.localeCompare(right)
      ).map(([name, allows]) => {
        const keyed = new Map(
          allows.map((permission) => {
            const atom = object(permission)!;
            const target = object(atom.target)!;
            const kind = String(target.kind);
            const key = kind === "apiSurface"
              ? [
                kind,
                target.api,
                target.surface,
                target.name,
                "",
                atom.action,
              ]
              : [
                kind,
                target.api,
                "operation",
                target.operation,
                target.signal,
                atom.action,
              ];
            return [key.join("\0"), permission] as const;
          }),
        );
        return [name, {
          allows: [...keyed.entries()].sort().map(([, value]) => value),
        }];
      }),
    );
  }

  for (const section of ["rpc", "operations", "events", "feeds", "state"]) {
    const definitions = object(contract[section]);
    if (!definitions || Object.keys(definitions).length === 0) continue;
    const lowered = structuredClone(definitions);
    for (const value of Object.values(lowered)) {
      const definition = object(value);
      if (!definition) continue;
      delete definition.subject;
      delete definition.capabilities;
      if (Array.isArray(definition.errors)) {
        definition.errors = [
          ...new Set(
            definition.errors.map((error) => object(error)?.type ?? error),
          ),
        ].sort();
        if (definition.errors.length === 0) delete definition.errors;
      }
      const transfer = object(definition.transfer);
      if (transfer) definition.transfer = { direction: transfer.direction };
      if (definition.internal === false) delete definition.internal;
      if (definition.cancel === false) delete definition.cancel;
      if (Object.keys(object(definition.signals) ?? {}).length === 0) {
        delete definition.signals;
      }
      if (Array.isArray(definition.params) && definition.params.length === 0) {
        delete definition.params;
      }
      if (definition.class === "domain") delete definition.class;
      if (
        Object.keys(object(definition.acceptedVersions) ?? {}).length === 0
      ) delete definition.acceptedVersions;
    }
    api[section] = lowered;
  }
  const errors = structuredClone(object(contract.errors) ?? {});
  for (const value of Object.values(errors)) {
    const definition = object(value);
    if (definition) delete definition.type;
  }
  for (const section of ["rpc", "operations"]) {
    for (const value of Object.values(object(api[section]) ?? {})) {
      const definition = object(value);
      const referenced = Array.isArray(definition?.errors)
        ? definition.errors
        : [];
      for (const error of referenced) {
        if (typeof error === "string" && !(error in errors)) errors[error] = {};
      }
    }
  }
  if (Object.keys(errors).length > 0) api.errors = errors;
  return api;
}

async function apiDigest(api: JsonObject): Promise<string> {
  const projection: JsonObject = { format: api.format, id: api.id };
  for (const field of ["schemas", "exports", "capabilities"]) {
    copy(api, projection, field);
  }
  for (
    const section of [
      "errors",
      "rpc",
      "operations",
      "events",
      "feeds",
      "state",
    ]
  ) {
    const definitions = object(api[section]);
    if (!definitions || Object.keys(definitions).length === 0) continue;
    const lowered = structuredClone(definitions);
    for (const value of Object.values(lowered)) {
      const definition = object(value);
      if (!definition) continue;
      delete definition.docs;
      if (section === "operations") {
        for (const signal of Object.values(object(definition.signals) ?? {})) {
          const signalDefinition = object(signal);
          if (signalDefinition) delete signalDefinition.docs;
        }
      }
    }
    projection[section] = lowered;
  }
  return (await digestJson(projection)).digest;
}

/**
 * Compile a contract manifest into canonical API and participant artifacts.
 * Referenced APIs are keyed by canonical API ID.
 */
export async function compileProtocolArtifacts(
  contract: ContractInput,
  referencedApis: Readonly<Record<string, JsonObject>> = {},
): Promise<CompiledProtocolArtifacts> {
  const source = sourceManifest(contract);
  const api = await compileApi(source, contract.CONTRACT_DIGEST);
  const apis: Record<string, JsonObject> = {
    ...referencedApis,
    [String(api.id)]: api,
  };
  const discoveredApis: JsonObject[] = [];
  for (const selected of contract[CONTRACT_RUNTIME]?.actions ?? []) {
    const dependencySource = actionSource(selected.action);
    if (!dependencySource) continue;
    const dependencyApi = await compileReferencedApi(dependencySource);
    apis[String(dependencyApi.id)] = dependencyApi;
    discoveredApis.push(dependencyApi);
  }
  const participant: JsonObject = { format: "trellis.participant.v1" };
  for (
    const field of [
      "id",
      "displayName",
      "description",
      "docs",
      "kind",
      "schemas",
    ]
  ) copy(source, participant, field);
  Object.values(object(participant.schemas) ?? {}).forEach(normalizeSchema);
  if (
    ["rpc", "operations", "events", "feeds", "state"].some((section) =>
      Object.keys(object(source[section]) ?? {}).length > 0
    )
  ) {
    const operationTransfers: JsonObject = {};
    for (
      const [name, value] of Object.entries(object(source.operations) ?? {})
    ) {
      const transfer = object(object(value)?.transfer);
      if (transfer?.direction !== "send") continue;
      const mapping = structuredClone(transfer);
      delete mapping.direction;
      operationTransfers[name] = mapping;
    }
    participant.implements = {
      self: {
        api: api.id,
        apiDigest: await apiDigest(api),
        ...(Object.keys(operationTransfers).length > 0
          ? { operationTransfers }
          : {}),
      },
    };
  }

  const contractUses = object(source.uses);
  const uses: JsonObject = {};
  for (const group of ["required", "optional"]) {
    const references = object(contractUses?.[group]);
    if (!references) continue;
    const lowered: JsonObject = {};
    for (const [alias, value] of Object.entries(references)) {
      const reference = object(value);
      const apiId = reference?.contract;
      if (typeof apiId !== "string" || !apis[apiId]) {
        throw new Error(
          `Referenced API artifact '${String(apiId)}' is required`,
        );
      }
      const referencedApi = apis[apiId];
      const used: JsonObject = {
        api: apiId,
        apiDigest: await apiDigest(referencedApi),
      };
      copy(reference!, used, "rpc");
      const operations = object(reference!.operations);
      if (operations) {
        const calls = Array.isArray(operations.call) ? operations.call : [];
        const cancels = Array.isArray(operations.cancel)
          ? operations.cancel
          : [];
        const controls = Array.isArray(operations.control)
          ? operations.control
          : [];
        const definitions = object(referencedApi.operations) ?? {};
        used.operations = {
          invoke: calls,
          observe: calls,
          cancel: cancels.filter((name) =>
            typeof name === "string" &&
            object(definitions[name])?.cancel === true
          ),
          ...(controls.length > 0
            ? {
              control: Object.fromEntries(controls.map((name) => [
                name,
                Object.keys(object(definitions[String(name)])?.signals ?? {})
                  .sort(),
              ])),
            }
            : {}),
        };
      }
      copy(reference!, used, "events");
      copy(reference!, used, "feeds");
      lowered[alias] = used;
    }
    if (Object.keys(lowered).length > 0) uses[group] = lowered;
  }
  if (Object.keys(uses).length > 0) participant.uses = uses;
  copy(source, participant, "resources");
  const jobs = object(source.jobs);
  if (jobs) {
    participant.jobQueues = structuredClone(object(jobs.queues) ?? jobs);
  }
  const consumers = object(source.eventConsumers);
  if (consumers) {
    participant.eventConsumers = Object.fromEntries(
      Object.entries(consumers).map(([name, value]) => {
        const consumer = structuredClone(object(value) ?? {});
        const events = structuredClone(object(consumer.uses) ?? {});
        if (consumer.self !== undefined) events.self = consumer.self;
        delete consumer.uses;
        delete consumer.self;
        consumer.events = events;
        return [name, consumer];
      }),
    );
  }
  return { api, participant, referencedApis: discoveredApis };
}

File: rust/crates/contracts/src/protocol_artifacts.rs
use std::collections::BTreeMap;

use serde_json::{json, Map, Value};
use trellis_protocol::{
    parse_api_v1, parse_participant_v1, resolve_participant_v1, ApiArtifactV1, GrantSetV1,
};

use crate::{digest_contract_value, parse_manifest, ContractsError};

/// Canonical protocol artifacts compiled from one contract authoring manifest.
#[derive(Clone, Debug)]
pub struct CompiledProtocolArtifacts {
    /// Canonical `trellis.api.v1` artifact owned by the contract.
    pub api: Value,
    /// Canonical `trellis.participant.v1` artifact for the runtime participant.
    pub participant: Value,
    /// Participant artifact digest used at bootstrap and authorization boundaries.
    pub participant_digest: String,
    /// Canonical participant needs digest.
    pub participant_needs_digest: String,
    /// Exact required permission atoms derived from selected API surfaces.
    pub required_grants: GrantSetV1,
    /// Exact optional permission atoms derived from selected API surfaces.
    pub optional_grants: GrantSetV1,
}

/// Compile contract authoring JSON into canonical API and participant artifacts.
///
/// `referenced_apis` is keyed by API ID and must contain every required or optional
/// contract reference selected by the participant.
///
/// # Errors
///
/// Returns an error when the contract, a referenced API, or the compiled protocol
/// artifacts fail validation or contextual resolution.
pub fn compile_protocol_artifacts(
    contract: &Value,
    referenced_apis: &BTreeMap<String, Value>,
) -> Result<CompiledProtocolArtifacts, ContractsError> {
    parse_manifest(contract.clone())?;
    let object = contract
        .as_object()
        .ok_or_else(|| invalid("contract must be an object".to_owned()))?;
    let contract_digest = digest_contract_value(contract)?;
    let api_value = compile_api(object, &contract_digest)?;
    let api = parse_api_v1(&api_value).map_err(|error| invalid(error.to_string()))?;
    let mut apis = BTreeMap::new();
    apis.insert(api.id().to_owned(), api.clone());
    for (id, value) in referenced_apis {
        let referenced = parse_api_v1(value).map_err(|error| invalid(error.to_string()))?;
        if referenced.id() != id {
            return Err(invalid(format!(
                "referenced API map key '{id}' does not match artifact id '{}'",
                referenced.id()
            )));
        }
        apis.insert(id.clone(), referenced);
    }
    let participant_value = compile_participant(object, &api, &apis)?;
    let participant =
        parse_participant_v1(&participant_value).map_err(|error| invalid(error.to_string()))?;
    let resolved =
        resolve_participant_v1(&participant, &apis).map_err(|error| invalid(error.to_string()))?;
    Ok(CompiledProtocolArtifacts {
        api: api
            .normalized_value()
            .map_err(|error| invalid(error.to_string()))?,
        participant: participant
            .normalized_value()
            .map_err(|error| invalid(error.to_string()))?,
        participant_digest: resolved.participant_digest().to_owned(),
        participant_needs_digest: resolved
            .needs()
            .digest()
            .map_err(|error| invalid(error.to_string()))?,
        required_grants: resolved.proposal().required().grant_set().clone(),
        optional_grants: resolved.proposal().optional().grant_set().clone(),
    })
}

fn compile_api(
    contract: &Map<String, Value>,
    contract_digest: &str,
) -> Result<Value, ContractsError> {
    let mut api = Map::new();
    api.insert(
        "format".to_owned(),
        Value::String("trellis.api.v1".to_owned()),
    );
    for field in [
        "id",
        "displayName",
        "description",
        "docs",
        "schemas",
        "exports",
    ] {
        copy(contract, &mut api, field);
    }
    normalize_embedded_schemas(&mut api);
    let schemas = api
        .entry("schemas".to_owned())
        .or_insert_with(|| Value::Object(Map::new()))
        .as_object_mut()
        .expect("compiled schemas are an object");
    if schemas.contains_key("TrellisContractArtifactIdentity") {
        return Err(invalid(
            "schema name 'TrellisContractArtifactIdentity' is reserved by the protocol artifact compiler"
                .to_owned(),
        ));
    }
    schemas.insert(
        "TrellisContractArtifactIdentity".to_owned(),
        json!({"type": "string", "const": contract_digest}),
    );
    for section in ["rpc", "operations", "events", "feeds", "state"] {
        if let Some(Value::Object(definitions)) = contract.get(section) {
            let mut lowered = definitions.clone();
            for definition in lowered.values_mut().filter_map(Value::as_object_mut) {
                definition.remove("subject");
                definition.remove("capabilities");
                if let Some(Value::Array(errors)) = definition.get_mut("errors") {
                    for error in errors {
                        if let Some(error_type) = error.get("type").and_then(Value::as_str) {
                            *error = Value::String(error_type.to_owned());
                        }
                    }
                }
            }
            if !lowered.is_empty() {
                api.insert(section.to_owned(), Value::Object(lowered));
            }
        }
    }
    let capabilities = compile_capabilities(contract);
    if !capabilities.is_empty() {
        api.insert("capabilities".to_owned(), Value::Object(capabilities));
    }
    if let Some(Value::Object(definitions)) = contract.get("errors") {
        let mut lowered = definitions.clone();
        for definition in lowered.values_mut().filter_map(Value::as_object_mut) {
            definition.remove("type");
        }
        if !lowered.is_empty() {
            api.insert("errors".to_owned(), Value::Object(lowered));
        }
    }
    let referenced_errors = ["rpc", "operations"]
        .into_iter()
        .filter_map(|section| api.get(section).and_then(Value::as_object))
        .flat_map(|definitions| definitions.values())
        .filter_map(|definition| definition.get("errors").and_then(Value::as_array))
        .flatten()
        .filter_map(Value::as_str)
        .map(str::to_owned)
        .collect::<Vec<_>>();
    if !referenced_errors.is_empty() {
        let errors = api
            .entry("errors".to_owned())
            .or_insert_with(|| Value::Object(Map::new()))
            .as_object_mut()
            .expect("compiled errors are an object");
        for error in referenced_errors {
            errors.entry(error).or_insert_with(|| json!({}));
        }
    }
    Ok(Value::Object(api))
}

fn compile_capabilities(contract: &Map<String, Value>) -> Map<String, Value> {
    let api_id = contract
        .get("id")
        .and_then(Value::as_str)
        .unwrap_or_default();
    let mut allows = BTreeMap::<String, Vec<Value>>::new();
    for (section, directions, surface) in [
        ("rpc", &[("call", "call")][..], "rpc"),
        (
            "operations",
            &[
                ("call", "invoke"),
                ("observe", "observe"),
                ("cancel", "cancel"),
            ][..],
            "operation",
        ),
        (
            "events",
            &[("publish", "publish"), ("subscribe", "subscribe")][..],
            "event",
        ),
        ("feeds", &[("subscribe", "subscribe")][..], "feed"),
    ] {
        let Some(definitions) = contract.get(section).and_then(Value::as_object) else {
            continue;
        };
        let mut names = definitions.keys().collect::<Vec<_>>();
        names.sort();
        for name in names {
            let definition = &definitions[name];
            let capabilities = definition.get("capabilities").and_then(Value::as_object);
            for (direction, action) in directions {
                for capability in capabilities
                    .and_then(|value| value.get(*direction))
                    .and_then(Value::as_array)
                    .into_iter()
                    .flatten()
                    .filter_map(Value::as_str)
                {
                    allows.entry(capability.to_owned()).or_default().push(json!({
                        "action": action,
                        "target": {"kind": "apiSurface", "api": api_id, "surface": surface, "name": name}
                    }));
                }
            }
            if section == "operations" {
                let mut signals = definition
                    .get("signals")
                    .and_then(Value::as_object)
                    .map(|value| value.keys().collect::<Vec<_>>())
                    .unwrap_or_default();
                signals.sort();
                for capability in capabilities
                    .and_then(|value| value.get("control"))
                    .and_then(Value::as_array)
                    .into_iter()
                    .flatten()
                    .filter_map(Value::as_str)
                {
                    for signal in &signals {
                        allows.entry(capability.to_owned()).or_default().push(json!({
                            "action": "control",
                            "target": {"kind": "operationSignal", "api": api_id, "operation": name, "signal": signal}
                        }));
                    }
                }
            }
        }
    }
    allows
        .into_iter()
        .map(|(name, allows)| (name, json!({"allows": allows})))
        .collect()
}

fn compile_participant(
    contract: &Map<String, Value>,
    own_api: &ApiArtifactV1,
    apis: &BTreeMap<String, ApiArtifactV1>,
) -> Result<Value, ContractsError> {
    let mut participant = Map::new();
    participant.insert(
        "format".to_owned(),
        Value::String("trellis.participant.v1".to_owned()),
    );
    for field in [
        "id",
        "displayName",
        "description",
        "docs",
        "kind",
        "schemas",
    ] {
        copy(contract, &mut participant, field);
    }
    normalize_embedded_schemas(&mut participant);
    let api_digest = own_api
        .digest()
        .map_err(|error| invalid(error.to_string()))?;
    if ["rpc", "operations", "events", "feeds", "state"]
        .iter()
        .any(|section| contract.get(*section).is_some_and(nonempty_object))
    {
        let operation_transfers = contract
            .get("operations")
            .and_then(Value::as_object)
            .into_iter()
            .flatten()
            .filter_map(|(name, operation)| {
                let mut transfer = operation.get("transfer")?.as_object()?.clone();
                (transfer.get("direction").and_then(Value::as_str) == Some("send")).then(|| {
                    transfer.remove("direction");
                    (name.clone(), Value::Object(transfer))
                })
            })
            .collect::<Map<_, _>>();
        let mut implementation = json!({"api": own_api.id(), "apiDigest": api_digest});
        if !operation_transfers.is_empty() {
            implementation["operationTransfers"] = Value::Object(operation_transfers);
        }
        participant.insert("implements".to_owned(), json!({"self": implementation}));
    }
    if let Some(Value::Object(groups)) = contract.get("uses") {
        let mut uses = Map::new();
        for group in ["required", "optional"] {
            let Some(Value::Object(references)) = groups.get(group) else {
                continue;
            };
            let mut lowered = Map::new();
            for (alias, reference) in references {
                let reference = reference
                    .as_object()
                    .ok_or_else(|| invalid(format!("uses.{group}.{alias} must be an object")))?;
                let api_id = reference
                    .get("contract")
                    .and_then(Value::as_str)
                    .ok_or_else(|| invalid(format!("uses.{group}.{alias}.contract is required")))?;
                let api = apis.get(api_id).ok_or_else(|| {
                    invalid(format!("referenced API artifact '{api_id}' is required"))
                })?;
                let mut used = Map::new();
                used.insert("api".to_owned(), Value::String(api_id.to_owned()));
                used.insert(
                    "apiDigest".to_owned(),
                    Value::String(api.digest().map_err(|error| invalid(error.to_string()))?),
                );
                copy(reference, &mut used, "rpc");
                if let Some(Value::Object(operations)) = reference.get("operations") {
                    let calls = operations.get("call").cloned().unwrap_or_else(|| json!([]));
                    let api_value = api
                        .normalized_value()
                        .map_err(|error| invalid(error.to_string()))?;
                    let cancel = operations
                        .get("cancel")
                        .and_then(Value::as_array)
                        .into_iter()
                        .flatten()
                        .filter(|name| {
                            name.as_str().is_some_and(|name| {
                                api_value["operations"][name]["cancel"].as_bool() == Some(true)
                            })
                        })
                        .cloned()
                        .collect::<Vec<_>>();
                    let control = operations
                        .get("control")
                        .and_then(Value::as_array)
                        .into_iter()
                        .flatten()
                        .filter_map(Value::as_str)
                        .map(|name| {
                            let signals = api_value["operations"][name]["signals"]
                                .as_object()
                                .map(|signals| {
                                    let mut names = signals.keys().cloned().collect::<Vec<_>>();
                                    names.sort();
                                    names
                                })
                                .unwrap_or_default();
                            (name.to_owned(), json!(signals))
                        })
                        .collect::<Map<_, _>>();
                    used.insert(
                        "operations".to_owned(),
                        json!({
                            "invoke": calls.clone(),
                            "observe": calls,
                            "cancel": cancel,
                            "control": control,
                        }),
                    );
                }
                copy(reference, &mut used, "events");
                copy(reference, &mut used, "feeds");
                lowered.insert(alias.clone(), Value::Object(used));
            }
            if !lowered.is_empty() {
                uses.insert(group.to_owned(), Value::Object(lowered));
            }
        }
        if !uses.is_empty() {
            participant.insert("uses".to_owned(), Value::Object(uses));
        }
    }
    if let Some(resources) = contract.get("resources") {
        participant.insert("resources".to_owned(), resources.clone());
    }
    if let Some(Value::Object(jobs)) = contract.get("jobs") {
        let queues = jobs
            .get("queues")
            .and_then(Value::as_object)
            .cloned()
            .unwrap_or_else(|| jobs.clone());
        participant.insert("jobQueues".to_owned(), Value::Object(queues));
    }
    if let Some(Value::Object(consumers)) = contract.get("eventConsumers") {
        let consumers = consumers
            .iter()
            .map(|(name, consumer)| {
                let mut consumer = consumer
                    .as_object()
                    .cloned()
                    .ok_or_else(|| invalid(format!("eventConsumers.{name} must be an object")))?;
                let mut events = consumer
                    .remove("uses")
                    .and_then(|value| value.as_object().cloned())
                    .unwrap_or_default();
                if let Some(owned) = consumer.remove("self") {
                    events.insert("self".to_owned(), owned);
                }
                consumer.insert("events".to_owned(), Value::Object(events));
                Ok((name.clone(), Value::Object(consumer)))
            })
            .collect::<Result<Map<_, _>, ContractsError>>()?;
        participant.insert("eventConsumers".to_owned(), Value::Object(consumers));
    }
    Ok(Value::Object(participant))
}

fn copy(source: &Map<String, Value>, target: &mut Map<String, Value>, key: &str) {
    if let Some(value) = source.get(key) {
        if !value.is_null() && !matches!(value, Value::Object(map) if map.is_empty()) {
            target.insert(key.to_owned(), value.clone());
        }
    }
}

fn nonempty_object(value: &Value) -> bool {
    value.as_object().is_some_and(|object| !object.is_empty())
}

fn normalize_embedded_schemas(artifact: &mut Map<String, Value>) {
    if let Some(Value::Object(schemas)) = artifact.get_mut("schemas") {
        for schema in schemas.values_mut() {
            normalize_schema(schema);
        }
    }
}

fn normalize_schema(value: &mut Value) {
    match value {
        Value::Array(values) => {
            for value in values {
                normalize_schema(value);
            }
        }
        Value::Object(object) => {
            if object.remove("patternProperties").is_some() {
                object.insert("additionalProperties".to_owned(), Value::Bool(true));
            }
            for value in object.values_mut() {
                normalize_schema(value);
            }
        }
        _ => {}
    }
}

fn invalid(details: String) -> ContractsError {
    ContractsError::SchemaValidation {
        kind: "compiled protocol artifacts",
        details,
    }
}
